go/merkletree/merkle_verifier.go

```diff
@@ -0,0 +1,176 @@
+package merkletree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+// MerkleVerifier is a class which knows how to verify merkle inclusion and consistency proofs.
+type MerkleVerifier struct {
+	treeHasher *TreeHasher
+}
+
+// NewMerkleVerifier returns a new MerkleVerifier for a tree based on the passed in hasher.
+func NewMerkleVerifier(h HasherFunc) MerkleVerifier {
+	return MerkleVerifier{
+		treeHasher: NewTreeHasher(h),
+	}
+}
+
+// VerifyInclusionProof verifies the correctness of the passed in proof given the passed in information about the tree and leaf.
+func (m MerkleVerifier) VerifyInclusionProof(leafIndex, treeSize int64, proof [][]byte, root []byte, data []byte) error {
+	calcRoot, err := m.RootFromInclusionProof(leafIndex, treeSize, proof, data)
+	if err != nil {
+		return err
+	}
+	if len(calcRoot) == 0 {
+		return errors.New("calculated empty root")
+	}
+	if !bytes.Equal(calcRoot, root) {
+		return fmt.Errorf("calculated root:\n%v\n does not match provided root:\n%s", calcRoot, root)
```

Comment:
A new error type might be called for here, so that callers can distinguish a bad inclusion proof from other errors without inspecting the error string (the latter are exceptional, the former is not).